### PR TITLE
Zy/38 rdflib navigation

### PR DESCRIPTION
**Pull request by zheyxu**
For issue https://github.com/anusii/rdflib/issues/38

### CODE REVIEW

File: lib/src/turtle/parsettl.dart [hunk 1]
```diff
@@ -0,0 +1,116 @@
+import '../graph.dart';
```
Comment (anushkavidanage) [inline lib/src/turtle/parsettl.dart:1]: Why do you have two functions in this file called `parseTTL` and `parseTTLStr`?

**Comment by gjwgit** [comment]
Thanks @anushkavidanage. Yes, we should only merge code that is required and cleanly implemented.

**Review by gjwgit — APPROVED** [review]
Approved pending fixing re @anushkavidanage comment.

File: lib/src/turtle/parsettl.dart (hunk 1 above)
Comment (zheyxu) [reply lib/src/turtle/parsettl.dart]: Removed.

**Review by gjwgit — APPROVED** [review]

File: lib/src/graph.dart [hunk 2]
```diff
@@ -270,27 +270,150 @@ class Graph {
...
+  /// (subject, predicate, or object) of the triples in the graph. It returns
+  /// a set of triples where any of these components matches the [value].
+  ///
+  /// Example usage:
+  /// ```dart
+  /// final value = 'exampleValue';
+  /// final matchingTriples = values(value);
+  /// print(matchingTriples);
+  /// ```
+  Set<Triple> tripleValues(String value) {
```
Comment (cdawei) [inline lib/src/graph.dart:398]: Perhaps naming it `triples()` instead of `tripleValues` if applicable.

File: lib/src/graph.dart [hunk 3]
```diff
@@ -270,27 +270,150 @@ class Graph {
   }
 
   /// Finds all subjects which have a certain predicate and object.
-  Set<URIRef> subjects(URIRef pre, dynamic obj) {
+  ///
+  /// If `pre` is provided, it checks for the predicate.
+  /// If `obj` is provided, it checks for the object.
+  /// If both are provided, it checks for both predicate and object.
+  /// If neither is provided, it returns all subjects in the triples.
+  Set<URIRef> subjects({URIRef? pre, dynamic obj}) {
```
Comment (cdawei) [inline lib/src/graph.dart:278]: Is it possible to simplify the code using Dart's *expr1 ?? expr2* [conditional expression](https://dart.dev/language/operators#conditional-expressions)?
For example, does the following cover the four cases you considered?
```
if (t.pre == (pre ?? t.pre) && t.obj == (obj ?? t.obj)) {
    objs.add(t.sub);
}
```
Similar changes might be applicable to the `predicates()` and `objects()` function.

File: lib/src/graph.dart (hunk 2 above)
Comment (zheyxu) [reply lib/src/graph.dart:398]: Hi @cdawei, the name 'triples' is already defined in the `parser/grammar_parser.dart`. The name can be renamed to 'matchTriples()'.

**Review by cdawei — APPROVED** [review]

**Review by anushkavidanage — APPROVED** [review]

**Pull request merged**